JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java

```diff
@@ -269,7 +301,51 @@ List<String> unifiedBulkCreateDerivations(List<List<String>> entitiesList, List<
 			}
 		}
 
-		modify.prefix(p_derived, p_agent);
+		// put sub query to retrieve the pure inputs whose timestamp is missing and add timestamp in insert clause
+		SubSelect sub = GraphPatterns.select();
+		Variable pureInput = sub.var();
+		Variable pureInputTimeInstant = sub.var();
+		Variable pureInputTimePosition = sub.var();
+		Variable existingTimestamp = sub.var();
+		Variable anyDerivation = sub.var();
+		long ts = Instant.now().getEpochSecond(); // get current epoch as timestamp
+		modify.insert(pureInput.has(hasTime, pureInputTimeInstant));
+		modify.insert(pureInputTimeInstant.isA(InstantClass).andHas(inTimePosition, pureInputTimePosition));
+		modify.insert(pureInputTimePosition.isA(TimePosition).andHas(numericPosition, ts).andHas(hasTRS, UnixTime));
+
+		// prepare values clause for the input timestamps
+		// should only add timestamp to potential pure inputs, i.e. those in inputsList and not in entitiesList
+		List<String> allOutputs = entitiesList.stream().flatMap(List::stream).collect(Collectors.toList());
+		List<String> potentialPureInputs = inputsList.stream().flatMap(List::stream).collect(Collectors.toList())
+				.stream().filter(e -> !allOutputs.contains(e)).collect(Collectors.toList());
+		// if the potentialPureInputs is empty, then means all inputs/entities cancels out
+		// this will lead to a circular dependency of the derivations to be created in bulk
+		if (potentialPureInputs.isEmpty()) {
+			throw new JPSRuntimeException(
+				"All inputs are cancelled out by outputs of derivation, resulting a circular dependency, inputs: " +
+				inputsList + "; outputs: " + entitiesList);
+		}
+		ValuesPattern pureInputTimestampVP = new ValuesPattern(pureInput, pureInputTimeInstant, pureInputTimePosition);
+		potentialPureInputs.stream().forEach(pureInputIri -> {
+			// create timestamp value pairs for the given entity
+			Iri time_instant_iri = iri(createTimeIRI());
```

Comment:
this is nitpicking, by Java conventions you shouldn't use underscores for variable names other than constants, I know this is also present in some of my old codes. Install sonarlint and it'll highlight all the issues

Reply:
This is changed as suggested in the latest commits.